### ISSUE TITLE
Deleted Devices still visible in Deconz

### ISSUE THREAD

**Issue by zigistar** [dresden-elektronik/deconz-rest-plugin#2064]
Hi,

some of my devices are still visible in Deconz despite they are removed 5 month ago and has been flagged as deleted in zll.db database (see screenshots). I also removed them multiple times in Deconz manually (Edit-Menu -> Delete selected Nodes). This removes the nodes only temporarily, they come back again when I reboot the Raspi.

What shall I do to make them disappear permanently? Shall I modify the zll.db with sqlite3?
Any recommendations?

Thanks a lot for your help.

I'm running 2.05.69 and firmware 26330500 on a Raspi 3+

![Bildschirmfoto 2019-11-11 um 22 14](https://user-images.githubusercontent.com/57197922/68622517-994fab80-04d2-11ea-8c1b-dd08fdf3d467.jpg)

![Bildschirmfoto 2019-11-11 um 22 12 40](https://user-images.githubusercontent.com/57197922/68622520-9b196f00-04d2-11ea-9bc6-64e3865a49f9.png)

**Comment by Pirol62** [comment]
Same issue but mine is connected to an osram plug 01. 
Not visible in Phoscon.

**Comment by ryancasler** [comment]
I have been having the same problem.  With 1 Tradfri Remote, 1 EcoLink Remote and 1 Lutron Connected Bulb remote.  No matter how many times i delete them they come right back.  Where else can I delete them from?  They have to be in a database table somewhere.

**Comment by realwax** [comment]
Same issue here... 
![image](https://user-images.githubusercontent.com/14924590/75201780-a8005c80-5769-11ea-95c6-474da8e0c6d8.png) As there is no active bulb or sensor numbering it can't be deleted via deconz rest api. Please fix!

**Comment by lensherm** [comment]
Has this been or is planning to be addressed?

**Comment by Buzztiger** [comment]
Same problem. They keep coming back after a restart. Anyone figured out how to remove them for good?

**Comment by ebaauw** [comment]
So one of my innr SP 120 plugs kept a Hue button end-device in its neighbour table, even though the button had left the network weeks ago.  Even after removing all references to the button from the database, deCONZ would re-create a node for the button, every time the plug's neighbour table was read.  Once you've ended up in a situation like this, there's nothing deCONZ can do.  I had to reset and re-pair the plug to remove the neighbour table entry.  Only after that, the button stopped re-appearing.

I want deCONZ to reset devices when they're removed (through the API or from the GUI), see #2377.  Note however, that won't prevent this situation from occurring ever again.  If the device's _Leave_ message won't reach one of its neighbours (e.g. because the neighbour is offline), it'll keep the device in the neighbour table entry.  Also, it might take up to an hour for the _Leave Request_ to reach a "deep sleeper" battery-powered end device.  If you reset or power down the device before that, it won't send the _Leave_ message.  You might want to wake battery-powered devices when removing them.

Until I'll have made the changes discussed in #2377, best remove devices through the API, supplying a `{"reset": true}` body to the DELETE request.

**Comment by meremortals70** [comment]
Another request here.  Need the ability to clean up devices when deleted.

**Comment by meremortals70** [comment]
The reason I opened a new issue was it would seem this was dead, is anyone working on this please?

**Comment by ebaauw** [comment]
Bump

**Comment by meremortals70** [comment]
Yep Deconz is no longer in my network

**Comment by ebaauw** [comment]
Bump

**Comment by realwax** [comment]
bump

**Comment by ebaauw** [comment]
@Mimiix, please assign the right label to shut up the bloody bot.

**Comment by labaland** [comment]
bumb

**Comment by sebres** [comment]
> Shall I modify the zll.db with sqlite3?

Just for the case someone will try to do such manual deletion from database (zll.db)...
After delete per SQL and restart of deCONZ, the device is still present, so I assume it must be still "unpaired" from conbee somehow ("Delete Node" obviously doesn't do that at the moment), so the gateway can forget all that devices.

I was unable to find in [User Manual deCONZ](https://www.dresden-elektronik.de/funk/software/deconz.html?file=files/dresden-elektronik/content/downloads/dokumente/funktechnik/deCONZ-BHB-en.pdf) how such removal can be done.
Unfortunately, I did not see this (and other) issue, as I chosen the gateway to purchase... 
Very questionable practice to ignore such serious issue (over 1 year)... Someone who gets just tested several devices is unable to "unregister" them, so it's opening a certain security vulnerability as well as probably can be affected by the limit - e. g. impossible to test, develop, whatever else with more than 200 devices once gets paired with self conbee stick.

Back to SQL, here is a statement that show all rows to be deleted:
```sql
.headers on
select * from sensors where sid in (select id from nodes where state = 'deleted');
select * from device_gui where device_id in (select id from nodes where state = 'deleted');
select * from devices where id in (select id from nodes where state = 'deleted');
select * from nodes where state = 'deleted';
```
Make a backup and execute the statement, after replacing of `select *` with `delete`... 

Just as already said, it won't help at all.

**Comment by ebaauw** [comment]
> After delete per SQL and restart of deCONZ, the device is still present, so I assume it must be still "unpaired" from conbee somehow ("Delete Node" obviously doesn't do that at the moment), so the gateway can forget all that devices.

That has nothing to do with pairing.  The deleted device would still be in the neighbour tables of its former neighbouring routers.  When the GUI queries these to draw the map, the node is "re-discovered".  Some devices seem to cleanup that table only when power cycled.  Best to delete devices only from the API, with `{"reset": true}`, as mentioned above.  Before resetting, the device sends a _Leave_ message, which causes its neighbouring routers to remove the device entry from their tables.

**Comment by sebres** [comment]
> That has nothing to do with pairing.

To me a delete imply everything (a complete removal so the gateway won't see that anymore), no matter how one would call that.

> Best to delete devices only from the API

For who it is better (why a default GUI or web-app cannot remove/unpair/whatever a device)?
Additionally doesn't API need to know what the device is (sensor, switch or light) to delete it with `{"reset": true}`? So what is about device which is a "leak" (neither in lights, nor sensors or switches). So then no chance without to reset gateway completely?

**Comment by Frank-Bemelman** [comment]
Hmm. Arrived in this thread today, to find there is not a solution yet. 

I opened the file /config/.storage/core.device_registry and removed the entries by hand. Saved file. Restarted HA and they were gone. 

Cheers,
Frank

**Comment by manup** [comment]
Deleted devices don't show up anymore since a while and are hidden / not loaded on startup.
